lib/propolis/src/enlightenment/hyperv/mod.rs

```diff
@@ -0,0 +1,313 @@
+                info!(
+                    &self.log,
+                    "established hypercall page at GPA {:#x}",
+                    new.gpa().0
+                );
+
+                WrmsrOutcome::Handled
+            } else {
+                info!(
+                    &self.log,
+                    "guest requested illegal hypercall GPA {:#x}",
+                    new.gpa().0
+                );
+                WrmsrOutcome::GpException
+            }
+        } else {
+            WrmsrOutcome::Handled
+        };
+
+        // Commit the new MSR value if the write was handled.
+        if outcome == WrmsrOutcome::Handled {
+            inner.msr_hypercall_value = new;
+        }
+
+        outcome
+    }
+}
+
+impl super::Enlightenment for HyperV {
+    fn add_cpuid(&self, cpuid: &mut CpuidSet) -> Result<(), AddCpuidError> {
+        let mut to_add = CpuidSet::new(cpuid.vendor());
+
+        let mut add_to_set = |id, val| {
+            to_add
+                .insert(id, val)
+                .expect("Hyper-V CPUID values don't conflict");
+        };
+
+        add_to_set(CpuidIdent::leaf(0x4000_0000), HYPERV_LEAF_0_VALUES);
+        add_to_set(CpuidIdent::leaf(0x4000_0001), HYPERV_LEAF_1_VALUES);
+        add_to_set(CpuidIdent::leaf(0x4000_0002), HYPERV_LEAF_2_VALUES);
+        add_to_set(CpuidIdent::leaf(0x4000_0004), HYPERV_LEAF_4_VALUES);
+        add_to_set(CpuidIdent::leaf(0x4000_0005), HYPERV_LEAF_5_VALUES);
+        add_to_set(CpuidIdent::leaf(0x4000_0006), HYPERV_LEAF_6_VALUES);
+        add_to_set(CpuidIdent::leaf(0x4000_0007), HYPERV_LEAF_7_VALUES);
+        add_to_set(CpuidIdent::leaf(0x4000_0008), HYPERV_LEAF_8_VALUES);
+        add_to_set(CpuidIdent::leaf(0x4000_0009), HYPERV_LEAF_9_VALUES);
+        add_to_set(CpuidIdent::leaf(0x4000_000A), HYPERV_LEAF_A_VALUES);
+
+        add_to_set(
+            CpuidIdent::leaf(0x4000_0003),
+            CpuidValues {
+                eax: HyperVLeaf3Eax::default().bits(),
+                ..Default::default()
+            },
+        );
+
+        super::add_cpuid(cpuid, to_add)
+    }
+
+    fn rdmsr(&self, vcpu: VcpuId, msr: MsrId) -> RdmsrOutcome {
+        match msr.0 {
+            HV_X64_MSR_GUEST_OS_ID => RdmsrOutcome::Handled(
+                self.inner.lock().unwrap().msr_guest_os_id_value,
+            ),
+            HV_X64_MSR_HYPERCALL => RdmsrOutcome::Handled(
+                self.inner.lock().unwrap().msr_hypercall_value.0,
+            ),
+            HV_X64_MSR_VP_INDEX => {
+                let id: u32 = vcpu.into();
+                RdmsrOutcome::Handled(id as u64)
+            }
+            _ => RdmsrOutcome::NotHandled,
+        }
+    }
+
+    fn wrmsr(&self, _vcpu: VcpuId, msr: MsrId, value: u64) -> WrmsrOutcome {
+        match msr.0 {
+            HV_X64_MSR_GUEST_OS_ID => self.handle_wrmsr_guest_os_id(value),
+            HV_X64_MSR_HYPERCALL => self.handle_wrmsr_hypercall(value),
+            HV_X64_MSR_VP_INDEX => WrmsrOutcome::GpException,
```

Comment:
since we're about to fault the guest, should we log this maybe?

Reply:
I'd be a little inclined to add a probe for this sort of thing at the rdmsr/wrmsr handler level in the vCPU code, but since this is indeed the defined outcome of trying to write to HV_X64_MSR_VP_INDEX I've left this alone for now.